core/store/src/adapter/trie_store.rs

```diff
@@ -100,18 +100,32 @@ impl<'a> TrieStoreUpdateAdapter<'a> {
         Self { store_update: StoreUpdateHolder::Reference(store_update) }
     }
 
+    /// Reads shard_uid mapping for given shard.
+    /// If the mapping does not exist, it means that `shard_uid` maps to itself.
+    fn read_shard_uid_mapping_from_db(
+        &self,
+        shard_uid: ShardUId,
+    ) -> Result<ShardUId, StorageError> {
+        let store = TrieStoreAdapter::new(Store::new(self.store_update.storage.clone()));
+        store.read_shard_uid_mapping_from_db(shard_uid)
+    }
+
     pub fn decrement_refcount_by(
         &mut self,
         shard_uid: ShardUId,
         hash: &CryptoHash,
         decrement: NonZero<u32>,
     ) {
-        let key = get_key_from_shard_uid_and_hash(shard_uid, hash);
+        let mapped_shard_uid =
+            self.read_shard_uid_mapping_from_db(shard_uid).expect("decrement_refcount_by failed");
+        let key = get_key_from_shard_uid_and_hash(mapped_shard_uid, hash);
```

Comment:
The `get_key_from_shard_uid_and_hash` method should only be used with mapped shard uid -> you can make that type safe by using the newtype pattern.

Reply:
Thanks, it helped with `entity_debug.rs`.